module-system/sov-modules-api/src/gas.rs

```diff
@@ -0,0 +1,190 @@
+pub mod sov_modules_api {
+    use anyhow::Result;
+
+    use core::marker::PhantomData;
+
+    // The `GasUnit` will be the responsible for scalar gas conversion, and will have dynamic
+    // types. We are expected to provide a couple of basic types on the SDK level, and the user
+    // might implement his own types
+
+    pub trait GasUnit {
+        type Price;
+
+        fn value(&self, price: &Self::Price) -> u64;
+    }
+
+    // The `Context` will be used to define the gas unit
+
+    pub trait Context {
+        type GasUnit: GasUnit;
+    }
+
+    pub struct DefaultContext<GU>(PhantomData<GU>)
+    where
+        GU: GasUnit;
+
+    impl<GU> Context for DefaultContext<GU>
+    where
+        GU: GasUnit,
+    {
+        type GasUnit = GU;
+    }
+
+    // We add `remainder_funds` to `WorkingSet` as a simple `u64`; the gas will already be converted
+    // to scalar value when charged from `WorkingSet`.
+
+    pub struct WorkingSet<C>
+    where
+        C: Context,
+    {
+        remainder_funds: u64,
+        _context: PhantomData<C>,
+    }
+
+    impl<C> WorkingSet<C>
+    where
+        C: Context,
+    {
+        pub fn new(remainder_funds: u64) -> Self {
+            Self {
+                remainder_funds,
```

Comment:
nit: let's call this `remaining_funds`. Remainder funds sounds like we're doing division